src/main/java/org/jenkinsci/plugins/workflow/support/pickles/ExecutorPickle.java

```diff
@@ -113,20 +114,27 @@ protected Executor tryResolve() throws Exception {
 
                 if (!future.isDone()) {
                     Queue.Task task = item.task;
-                    Label nodeNameLabel = task.getAssignedLabel();  // Once execution started, this was set to the Computer name
-                    // See javadocs for ExecutorStepExecution.PlaceholderTask.label
-                    long waitTime = System.currentTimeMillis()-startTimeMillis;
-                    // TODO see if we need to do something with the ExecutorStepExecution.PlaceholderTask.runningTasks
-                    if (nodeNameLabel != null && task.getLastBuiltOn() == null) {  // Can't find node it was built on!
-                        if (waitTime > TIMEOUT_WAITING_FOR_EPHMERAL_NODE && task instanceof ExecutorStepExecution.PlaceholderTask) {
-                            Queue.getInstance().cancel(item);
-                            throw new AbortException(MessageFormat.format("Killed {0} after waiting for {1} millis because we assume unknown Node {1} (ephemeral={2}) is never going to reappear!",
-                                    new Object[]{item, waitTime, nodeNameLabel.toString(), isEphemeral()}));
+                    if (task instanceof ExecutorStepExecution.PlaceholderTask) {
+                        ExecutorStepExecution.PlaceholderTask placeholder = (ExecutorStepExecution.PlaceholderTask)task;
+
+                        // Non-null cookie means body has begun execution, i.e. we started using a node
+                        // PlaceholderTask#getAssignedLabel is set to the Node name when execution starts
+                        // Thus we're guaranteeing the execution began and the Node is now unknown.
+                        // Theoretically it's safe to simply fail earlier when rehydrating any EphemeralNode... but we're being extra safe.
+                        if (placeholder.getCookie() != null && Jenkins.getActiveInstance().getNode(placeholder.getAssignedLabel().getName()) == null ) {
+                            if (isEphemeral() || System.nanoTime() > endTimeNanos) {
+                                Queue.getInstance().cancel(item);
+                                if (isEphemeral()) {
+                                    throw new AbortException(MessageFormat.format("Killed {0} because EphemeralNode {1} is never going to reappear, by definition!",
+                                            new Object[]{item, task.getAssignedLabel().toString()}));
```

Comment:
BTW `Queue.Item.toString()` is not going to be particularly user-friendly; will look something like

    hudson.model.Queue$WaitingItem:ExecutorStepExecution.PlaceholderTask{runId=folder/jobname#123,label=,context=CpsStepContext[23:node]:Owner[folder/jobname/123:folder/jobname #123],cookie=dead-beef-1234-5678}

Whew!